src/python/lib/ensembl/compara/citest/testdb.py

```diff
@@ -0,0 +1,194 @@
+            return excinfo.value.args[0] + "\n"
+        return super().repr_failure(excinfo, style)
+
+    def get_report_header(self) -> str:
+        """Returns the header to display in the error report."""
+        return f"Database table: {self.table}, test: {self.name}"
+
+    def test_num_rows(self, variation: float = 0.0, group_by: Union[str, List] = None,
+                      filter_by: Union[str, List] = None) -> None:
+        """Compares the number of rows between reference and target tables.
+
+        If `group_by` is provided, the number of rows will be compared per group, applying the same variation
+        to all of them. If `filter_by` is provided, only the rows matching all the given conditions will be
+        compared.
+
+        Args:
+            variation: Allowed variation between reference and target tables.
+            group_by: Group rows by column(s), and count the number of rows per group.
+            filter_by: Filter rows by one or more conditions (joined by the AND operator).
+
+        Raise:
+            FailedDBTestException: If `group_by` is provided and the number of groups is different; or if the
+                number of rows differ for at least one group.
+
+        """
+        # Compose the sql query from the given parameters (both databases should have the same table schema)
+        table = self.ref_db.tables[self.table]
+        columns = [table.columns[col] for col in to_list(group_by)]
+        # Use primary key in count to improve the query performance
+        primary_key = self.ref_db.get_primary_key_columns(self.table)[0]
+        query = select(columns + [func.count(table.columns[primary_key]).label('nrows')])
+        if columns:
+            # ORDER BY to ensure that the results are always in the same order (for the same groups)
+            query = query.group_by(*columns).order_by(*columns)
+        for clause in to_list(filter_by):
+            query = query.where(text(clause))
+        # Get the number of rows for both databases
+        ref_data = pandas.read_sql(query, self.ref_db.connect())
+        target_data = pandas.read_sql(query, self.target_db.connect())
+        # Check if the size of the returned tables are the same
+        if ref_data.shape != target_data.shape:
+            expected = ref_data.shape[0]
+            found = target_data.shape[0]
+            # Note that the shape can only be different if group_by is given
+            message = (
+                f"Different number of groups ({', '.join([c.name for c in columns])}) for table "
+                f"'{self.table}'"
+            )
+            raise FailedDBTestException(expected, found, query, message)
+        # Check if the number of rows (per group) are the same
+        difference = abs(ref_data['nrows'] - target_data['nrows'])
+        allowed_variation = ref_data['nrows'] * variation
+        failing_rows = difference > allowed_variation
+        if failing_rows.any():
```

Comment:
I would expect `failing_rows` to be a boolean

Reply:
`failed_rows`? `wrong_rows`?

Reply:
I mean: why `if failing_rows.any():` and not `if failing_rows:` ?

Reply:
Ah, I think I understand: `ref_data['nrows']` selects the column named `nrows`, so it's a list (or a pandas dataframe). I thought it was returning a single integer